### ISSUE TITLE
Writing lint: How to detect access to a local from a closure?

### ISSUE THREAD

**Issue by schubart** [rust-lang/rust-clippy#9387]
(I hope it's OK to open an issue to ask a questions. If not, please tell me where I should ask the question instead.)

I'm trying to find all usages of a local, but I fail to find cases where a local is accessed from a closure.

Simplified code that tries to find *any* access to *each* local:

```rust
use clippy_utils::diagnostics::span_lint;
use clippy_utils::get_enclosing_block;
use clippy_utils::path_to_local_id;
use if_chain::if_chain;
use rustc_hir::intravisit::walk_expr;
use rustc_hir::intravisit::Visitor;
use rustc_hir::Expr;
use rustc_hir::HirId;
use rustc_hir::Local;
use rustc_hir::PatKind;
use rustc_lint::LateContext;
use rustc_lint::LateLintPass;
use rustc_session::declare_lint_pass;
use rustc_session::declare_tool_lint;

declare_clippy_lint! {
    #[clippy::version = "1.64.0"]
    pub COLLECTION_IS_NEVER_READ,
    nursery,
    "default lint description"
}
declare_lint_pass!(CollectionIsNeverRead => [COLLECTION_IS_NEVER_READ]);

impl<'tcx> LateLintPass<'tcx> for CollectionIsNeverRead {
    fn check_local(&mut self, cx: &LateContext<'tcx>, local: &'tcx Local<'tcx>) {
        if_chain! {
            if let PatKind::Binding(_, local_id, _, _) = local.pat.kind;
            if let Some(enclosing_block) = get_enclosing_block(cx, local.hir_id);
            then {
                let mut visitor = MyVisitor {
                    cx,
                    local_id,
		};
                visitor.visit_block(enclosing_block);
            }
	}
    }
}

struct MyVisitor<'a, 'tcx> {
    cx: &'a LateContext<'tcx>,

    local_id: HirId,
}

impl<'a, 'tcx> Visitor<'tcx> for MyVisitor<'a, 'tcx> {
    fn visit_expr(&mut self, expr: &'tcx Expr<'_>) {
        if path_to_local_id(expr, self.local_id) {
            span_lint(self.cx, COLLECTION_IS_NEVER_READ, expr.span, "local is being used here");
        }
        walk_expr(self, expr);
    }
}
```

Test program:

```rust
#![warn(clippy::collection_is_never_read)]

fn main() {
    let x = 10;

    println!("first usage: {x}");

    let _ = { || println!("second usage: {x}") };

    println!("third usage: {x}");
}
```

My lint only finds "first usage" and "third usage".

How do I make it find "second usage"?

Thanks!

**Comment by llogiq** [comment]
You'll have to check closure bodies. `ExprKind::Closure(..)` has a body ID, which you can use to get the body (using cx.tcx, sorry, I'm on mobile right now), which you can then walk.

**Comment by lukaslueg** [comment]
https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/unused_peekable.rs might be of use

**Comment by schubart** [comment]
Thanks @llogiq that worked:

```rust
impl<'a, 'tcx> Visitor<'tcx> for MyVisitor<'a, 'tcx> {
    fn visit_expr(&mut self, expr: &'tcx Expr<'_>) {
        if path_to_local_id(expr, self.local_id) {
            span_lint(self.cx, COLLECTION_IS_NEVER_READ, expr.span, "local is being used here");
        }

        if let ExprKind::Closure(&Closure { body, .. }) = expr.kind {
            // We have to go deeper                                                                                                                                                                                             
            let body = self.cx.tcx.hir().body(body);
            self.visit_body(body);
        }

        walk_expr(self, expr);
    }
}
```

It seems a bit surprising that the visitor does not walk into closure bodies.

Thanks @lukaslueg for the pointer, but I failed to see how this code relates to my problem. Thanks anyhow!

**Comment by Jarcho** [comment]
Visitors don't enter bodies (e.g. closures, const blocks) or items by default. You can make them do so by setting the `NestedFilter` associated type appropriately. You can look at the `clippy_utils::visitors` module for various examples.

**Comment by schubart** [comment]
Thanks @Jarcho. `clippy_utils::visitors::expr_visitor` works well and I no longer have to implement `Visitor` myself:

```rust
use clippy_utils::diagnostics::span_lint;
use clippy_utils::get_enclosing_block;
use clippy_utils::path_to_local_id;
use clippy_utils::visitors::expr_visitor;
use if_chain::if_chain;
use rustc_hir::intravisit::Visitor;
use rustc_hir::Local;
use rustc_hir::PatKind;
use rustc_lint::LateContext;
use rustc_lint::LateLintPass;
use rustc_session::declare_lint_pass;
use rustc_session::declare_tool_lint;

declare_clippy_lint! {
    #[clippy::version = "1.64.0"]
    pub COLLECTION_IS_NEVER_READ,
    nursery,
    "default lint description"
}
declare_lint_pass!(CollectionIsNeverRead => [COLLECTION_IS_NEVER_READ]);

impl<'tcx> LateLintPass<'tcx> for CollectionIsNeverRead {
    fn check_local(&mut self, cx: &LateContext<'tcx>, local: &'tcx Local<'tcx>) {
        if_chain! {
            if let PatKind::Binding(_, local_id, _, _) = local.pat.kind;
            if let Some(enclosing_block) = get_enclosing_block(cx, local.hir_id);
            then {

                let mut visitor = expr_visitor(cx, |expr| {
                    if path_to_local_id(expr, local_id) {
                        span_lint(cx, COLLECTION_IS_NEVER_READ, expr.span, "local is being used here");
                    }

                    true // Keep going                                                                                                                                                                                          
                });

                visitor.visit_block(enclosing_block);
            }
        }
    }
}
```
